app/main/views/user_profile.py

```diff
@@ -420,14 +444,42 @@ def user_profile_2fa():
                 # Default to email auth if something unexpected is selected
                 auth_type = "email_auth"
 
+            # If the user is switching to SMS, ensure they have a mobile number
+            if auth_type == "sms_auth" and not current_user.mobile_number:
+                flash(_("You must add a mobile number to your profile to use this option."), "error")
+                session["from_send_page"] = "user_profile_2fa"
+                return redirect(url_for(".user_profile_mobile_number"))
+
             # Update the user's authentication type
             current_user.update(auth_type=auth_type)
 
             # Flash a success message
-            flash(_("Two-factor authentication method updated"), "default_with_tick")
+            flash(_("Two-step verification method updated"), "default_with_tick")
 
-            # Redirect back to user profile
+            # Redirect back to user profile and revoke their additional authentication
+            session.pop(HAS_AUTHENTICATED, None)
             return redirect(url_for(".user_profile"))
 
-        return render_template("views/user-profile/2fa.html", form=form, hints=hints)
+        return render_template("views/user-profile/2fa.html", form=form, hints=hints, badge_options=badge_options)
     return redirect(url_for(".user_profile"))
+
+
+@main.route("/user-profile/2fa/authenticate", methods=["GET", "POST"])
+@user_is_logged_in
+def user_profile_2fa_authenticate():
+    # Validate password for form
+    def _check_password(pwd):
+        return user_api_client.verify_password(current_user.id, pwd)
+
+    form = ConfirmPasswordForm(_check_password)
+
+    if form.validate_on_submit():
+        session[HAS_AUTHENTICATED] = True
+        return redirect(url_for(".user_profile_2fa"))
+
+    return render_template(
+        "views/user-profile/authenticate.html",
+        thing=_("email address"),
+        form=form,
+        back_link=url_for(".user_profile_email"),
```

Comment:
I think this should be `.user_profile`. Currently when you click the link to change 2FA then go back without entering the password, it goes to the add/change email address page. Clicking cancel sends you back to `.user_profile`.

Reply:
Makes sense, good catch!
```suggestion
        back_link=url_for(".user_profile"),
```